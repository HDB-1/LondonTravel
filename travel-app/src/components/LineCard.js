import React, { Component } from 'react'
import '../styling/LineCard.css'


export default class Items extends Component {

    render() {
        let { name, colour, status, reason } = this.props;
        return (
            <div className="card">
                <div className="card-header" style={colour}>
                    <h3>{name}</h3>
                </div>
                <div className="card-body">
                    <h5>{status}</h5>
                    {reason && <p>{reason}</p>}
                </div> 
                <div className="card-footer"><br/></div>
            </div>
        )
    }
}
