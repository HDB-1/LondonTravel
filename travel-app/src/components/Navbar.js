import React from 'react';
import '../styling/Navbar.css';
import 'bootstrap/dist/css/bootstrap.min.css';
import { Navbar, Nav } from 'react-bootstrap';

function NavBar() {
    return (
        <div>
            <Navbar bg="primary" expand="lg" variant="dark">
                <Navbar.Brand href="">London Travel</Navbar.Brand>
                <Navbar.Toggle aria-controls="basic-navbar-nav" />
                <Navbar.Collapse id="basic-navbar-nav">
                    <Nav className="mr-auto">
                        <Nav.Link className="link" href="/" id="home_link">Home</Nav.Link>
                        <Nav.Link className="link" href="/linestatus" id="line_link">Lines</Nav.Link>
                        <Nav.Link className="link" href="/map" id="map_link">Tube Map</Nav.Link>
                        <Nav.Link className="link" href="/weather" id="weather_link">Weather</Nav.Link>
                    </Nav>
                </Navbar.Collapse>
            </Navbar>
        </div>
    )
}

export default NavBar;