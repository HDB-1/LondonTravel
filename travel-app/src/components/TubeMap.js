import React from 'react';
import map from '../assets/tube_map.png'
import '../styling/TubeMap.css';

export default function TubeMap() {
    return (
        <div>
            <img src={map} />
        </div>
    )
}
