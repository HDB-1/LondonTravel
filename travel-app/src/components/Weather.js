import React, { Component } from 'react';
import '../styling/Weather.css';

export default class Weather extends Component {

    bearings = (wind_degree) => {
        if(wind_degree > 335 && wind_degree < 26) {
            return "Northerly"
        } else if(wind_degree > 25 && wind_degree < 65) {
            return "North Easterly"
        } else if(wind_degree > 65 && wind_degree < 115) {
            return "Easterly"
        } else if(wind_degree > 115 && wind_degree < 155) {
            return "South Easterly"
        } else if(wind_degree > 155 && wind_degree < 205) {
            return "Southerly"
        } else if(wind_degree > 205 && wind_degree < 245) {
            return "South Westerly"
        } else if(wind_degree > 245 && wind_degree < 295) {
            return "Westerly"
        } else if(wind_degree > 295 && wind_degree < 335) {
            return "North Westerly"
        }
      }
    componentDidMount() {
        this.bearings()
    }

    render() {
        let { temp_c, feelslike_c, wind_mph, wind_degree, condition } = this.props;
        let direction = this.bearings(wind_degree)
        return (
            <div className="container">
                <div className="weather-header">
                    <h2>Weather</h2>
                </div>
                <div className="weather-body">
                    <h3>{condition}</h3>
                    <h5>Temp: {temp_c}</h5>
                    <h5>Feels like: {feelslike_c}</h5>
                    <h5>Wind: {wind_mph}</h5>
                    <h5>Wind direction: {direction}</h5>
                </div> 
                <div className="weather-footer">
                    <p>Powered by <a href="https://www.weatherapi.com/" title="Free Weather API">WeatherAPI.com</a></p>
                </div>
            </div> 
        )
    }
}
