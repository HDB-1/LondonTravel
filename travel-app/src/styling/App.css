@import url('https://fonts.googleapis.com/css?family=Quicksand&display=swap');

h1 {
  text-align: center;
  font-size: 50px;
  font-family: 'Quicksand';
  text-decoration: underline;
}

.flex-container {
  display: flex;
  justify-content: center;
  flex-direction: row;
  flex-wrap: wrap;
  width: 100%;
  height: 100%;
  padding: 5px;
}

div[class="homepage"] {
  display: flex;
  justify-content: center;
  flex-direction: row;
  flex-wrap: wrap;
  margin-top: 5%;

}
h4 {
  font-variant: small-caps;
  text-align: center;
  font-size: 30px;
  font-family: 'Quicksand';
}