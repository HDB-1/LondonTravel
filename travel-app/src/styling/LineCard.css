div[class="card"] {
    display: flex;
    flex-direction: column;
    justify-content: center;
    margin-right: 10px;
    margin-bottom: 10px;
    width: 400px;
    border: 2px solid black;
}

div[class="card-header"] {
    text-align: center;
}