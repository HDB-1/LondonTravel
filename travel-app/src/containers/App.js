import React, { Component } from 'react'
import Axios from 'axios';
import { BrowserRouter as Router , Switch, Route } from 'react-router-dom';
import Navbar from '../components/Navbar';
import LineCard from '../components/LineCard';
import TubeMap from '../components/TubeMap';
import '../styling/App.css';
import 'bootstrap/dist/css/bootstrap.min.css';

export default class App extends Component {

  constructor(props) {
    super(props);
    this.state = {
      fetched: false,
      lineData: [],
      lineColour: {
        bakerloo: '#894E24',
        central: '#DC251F',
        circle: '#F8CE00',
        district: '#197229',
        "hammersmith-city": "#D799AF",
        jubilee: "#6A7278",
        metropolitan: "#751056",
        northern: "#000000",
        piccadilly: "#2119A8",
        victoria: "#3EA0E2",
        "waterloo-city": "#76D0BD",
        "london-overground": "#E86A11",
        dlr: '#38AFAD',
        tram: "#66CC01",
      }
    }
  }

  getAllData = () => {
    Axios.get("https://api.tfl.gov.uk/Line/Mode/tube%2Cdlr%2Coverground%2Ctram/Status")
    .then(res => {
      this.setState({
        lineData: res.data,
        fetched: true
      })
    })
    .catch(err => {
      console.log(err)
    })
  }

  componentDidMount() {
    this.getAllData()
  }

  render() {
    let { lineData, lineColour } = this.state;
    return (
      <Router>
        <Navbar/>
        <Switch>
          <Route exact path="/">
            <p>Welcome to the app, get London travel info</p>
          </Route>
          <Route exact path="/linestatus">
            <div className="flex-container">
            {lineData.map(line => {
              let colour = { backgroundColor: lineColour[line.id] }
              return (<LineCard
                name={line.name}
                colour={colour}
                status={line.lineStatuses[0]["statusSeverityDescription"]}
                reason={line.lineStatuses[0]["reason"]} />)})}
                </div>
          </Route>
          <Route exact path="/map">
            <TubeMap />
          </Route>
        </Switch>
      </Router>
    )
  }
}

