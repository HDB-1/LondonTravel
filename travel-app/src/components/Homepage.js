import React from 'react'
import '../styling/App.css';

export default function Homepage() {
    return (
        <div className="homepage">
            <h4>HDB's London Travel App</h4>
        </div>
    )
}
