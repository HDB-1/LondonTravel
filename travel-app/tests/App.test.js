import React from 'react';
import { shallow, mount } from 'enzyme';
import { MemoryRouter, Link } from "react-router-dom";
import App from '../src/containers/App';
import Navbar from '../src/components/Navbar';
import LineCard from '../src/components/LineCard';
import TubeMap from '../src/components/TubeMap';
import Weather from '../src/components/Weather';
import Homepage from '../src/components/Homepage';
import NotFound from '../src/components/NotFound';
import waitUntil from "async-wait-until";

describe("shallow app", () => {
    let wrapper;

    beforeEach(() => {
        wrapper = shallow(<App />)
    })
    it("should render correctly", () => {
        expect(wrapper).toMatchSnapshot();
    })
    it("should contain a Navbar Component", () => {
        expect(wrapper.containsMatchingElement(<Navbar />)).toEqual(true);
    })
    // it("should contain a LineCard Component", () => {
    //     expect(wrapper.containsMatchingElement(<LineCard />)).toEqual(true);
    // })
    it("should contain a TubeMap Component", () => {
        expect(wrapper.containsMatchingElement(<TubeMap />)).toEqual(true);
    })
    it("should contain a Weather Component", () => {
        expect(wrapper.containsMatchingElement(<Weather />)).toEqual(true);
    })
})

describe("Mounted App", () => {
    it("Should render a <Navar /> component on any route", () => {
        let wrapper = mount(
          <MemoryRouter initialEntries={["/"]}>
            <App />
          </MemoryRouter>
        );
        expect(wrapper.containsMatchingElement(<Navbar />)).toEqual(true);
        wrapper = mount(
          <MemoryRouter initialEntries={["/linestatus"]}>
            <App />
          </MemoryRouter>
        );
        expect(wrapper.containsMatchingElement(<Navbar />)).toEqual(true);
        wrapper = mount(
          <MemoryRouter initialEntries={["/map"]}>
            <App />
          </MemoryRouter>
        );
        expect(wrapper.containsMatchingElement(<Navbar />)).toEqual(true);
        wrapper = mount(
          <MemoryRouter initialEntries={["/weather"]}>
            <App />
          </MemoryRouter>
        );
        expect(wrapper.containsMatchingElement(<Navbar />)).toEqual(true);
      });
      it("Should render a <NotFound /> component on any route not covered by our specific routing layout", () => {
        let wrapper = mount(
          <MemoryRouter initialEntries={["/random"]}>
          <App />
        </MemoryRouter>
        );
        expect(wrapper.containsMatchingElement(<LineCard />)).toEqual(false);
        expect(wrapper.containsMatchingElement(<TubeMap />)).toEqual(false);
        expect(wrapper.containsMatchingElement(<Weather />)).toEqual(false);
        expect(wrapper.containsMatchingElement(<NotFound />)).toEqual(true);
      })
    
    //   it("Navbar should send user to the homepage when click Home", () => {
    //     let wrapper = mount(
    //       <MemoryRouter initialEntries={["/weather"]}>
    //         <App />
    //       </MemoryRouter>
    //     );
    //     expect(wrapper.containsMatchingElement(<Weather />)).toEqual(true);
    //     expect(wrapper.containsMatchingElement(<Homepage />)).toEqual(false);
    //     wrapper
    //         .find(<Navbar />)
    //         .find("Nav.Link")
    //         .find("#home_link")
    //         .simulate("click", { button: 0 });
    //     expect(wrapper.containsMatchingElement(<Weather />)).toEqual(false);
    //     expect(wrapper.containsMatchingElement(<Homepage />)).toEqual(true);
    //   });
});

// <========================== API TESTING =============================>

function weatherAPIRequest () {
    let url = "http://api.weatherapi.com/v1/current.json?key=a8c23d3c2d0a43d78c5172954200303&q=london";
    return fetch(url).then(res => res.json());
  };
function linestatusAPIRequest () {
    let url = "https://api.tfl.gov.uk/Line/Mode/tube%2Cdlr%2Coverground%2Ctram/Status";
    return fetch(url).then(res => res.json());
  };

describe("testing API", () => {
    beforeEach(() => {
      fetch.resetMocks();
    });
    it("calls API and returns data to me", () => {
      fetch.mockResponseOnce(JSON.stringify({ data: "placeholder12345" }));
      linestatusAPIRequest().then(res => {
        expect(res.data).toEqual("placeholder12345");
      });
      expect(fetch.mock.calls.length).toEqual(1);
      expect(fetch.mock.calls[0][0]).toEqual(
        'https://api.tfl.gov.uk/Line/Mode/tube%2Cdlr%2Coverground%2Ctram/Status'
      );
    });
    it("calls API and returns data to me", () => {
      fetch.mockResponseOnce(JSON.stringify({ data: "placeholder12345" }));
      weatherAPIRequest().then(res => {
        expect(res.data).toEqual("placeholder12345");
      });
      expect(fetch.mock.calls.length).toEqual(1);
      expect(fetch.mock.calls[0][0]).toEqual(
        'http://api.weatherapi.com/v1/current.json?key=a8c23d3c2d0a43d78c5172954200303&q=london'
      );
    });
  });
