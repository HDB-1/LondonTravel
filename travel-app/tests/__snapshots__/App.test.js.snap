// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shallow app should render correctly 1`] = `
<Component>
  <NavBar />
  <Switch>
    <Route
      exact={true}
      path="/"
    >
      <Homepage />
    </Route>
    <Route
      exact={true}
      path="/linestatus"
    >
      <div
        className="flex-container"
      />
    </Route>
    <Route
      exact={true}
      path="/map"
    >
      <TubeMap />
    </Route>
    <Route
      exact={true}
      path="/weather"
    >
      <Weather />
    </Route>
    <Route>
      <NotFound />
    </Route>
  </Switch>
</Component>
`;
